Documentation/contributing/development/hive.rst

```diff
@@ -0,0 +1,797 @@
+functions are not yet executed as our application may have multiple hives
+and we need to delay object instantiation to until we know which hive to use.
+
+However ``New`` does execute an invoke function to gather all command-line flags from
+all configuration cells. These can be then registered with a Cobra command:
+
+.. code-block:: go
+
+    var cmd *cobra.Command = /* ... */
+    myHive.RegisterFlags(cmd.Flags())
+
+After that the hive can be started with ``myHive.Run()``.
+
+Run() will first construct the parsed configurations and will then execute
+all invoke functions to instantiate all needed objects. As part of this the
+lifecycle hooks will have been appended (in dependency order). After that
+the start hooks can be executed one after the other to start the hive. Once
+started, Run() waits for SIGTERM and SIGINT signals and upon receiving one
+will execute the stop hooks in reverse order to bring the hive down.
+
+Now would be a good time to try this out in practice. You'll find a small example
+application in `pkg/hive/example <https://github.com/cilium/cilium/tree/master/pkg/hive/example>`_.
+Try running it with ``go run .`` and exploring the implementation (try what happens if a provider is commented out!).
+
+Inspecting a hive
+~~~~~~~~~~~~~~~~~
+
+The ``hive.Hive`` can be inspected with the 'hive' command after it's
+been registered with cobra:
+
+.. code-block:: go
+
+    var rootCmd *cobra.Command = /* ... */
+    rootCmd.AddCommand(myHive.Command())
+
+.. code-block:: shell-session
+
+    cilium$ go run ./daemon hive
+    Cells:
+
+    Ⓜ️ agent (Cilium Agent):
+      Ⓜ️ infra (Infrastructure):
+        Ⓜ️ k8s-client (Kubernetes Client):
+             ⚙️ (client.Config) {
+                 K8sAPIServer: (string) "",
+                 K8sKubeConfigPath: (string) "",
+                 K8sClientQPS: (float32) 0,
+                 K8sClientBurst: (int) 0,
+                 K8sHeartbeatTimeout: (time.Duration) 30s,
+                 EnableK8sAPIDiscovery: (bool) false
+             }
+ 
+             🚧 client.newClientset (cell.go:109):
+                 ⇨ client.Config, hive.Lifecycle, logrus.FieldLogger 
+                 ⇦ client.Clientset 
+    ...
+
+    Start hooks:
+
+        • gops.registerGopsHooks.func1 (cell.go:44)
+        • cmd.newDatapath.func1 (daemon_main.go:1625)
+        ...
+
+    Stop hooks:
+        ...
+   
+
+The hive command prints out the cells, showing what modules, providers,
+configurations etc. exist and what they're requiring and providing.
+Finally the command prints out all registered start and stop hooks.
+Note that these hooks often depend on the configuration (e.g. k8s-client
+will not insert a hook unless e.g. --k8s-kubeconfig-path is given). The
+hive command takes the same command-line flags as the root command.
+
+The provider dependencies in a hive can also be visualized as a graphviz dot-graph:
+
+.. code-block:: bash
+
+    cilium$ go run ./daemon hive dot-graph | dot -Tx11
+
+Guidelines
```

Comment:
These deserve an introduction sentence. Maybe even call out specifically that these can and should be used as Code review comments like https://github.com/golang/go/wiki/CodeReviewComments

Reply:
:+1: